aten/src/ATen/core/dispatch/DispatchKeyExtractor.h

```diff
@@ -0,0 +1,85 @@
+#pragma once
+
+#include <cstdint>
+#include <ATen/core/function_schema.h>
+#include <ATen/core/jit_type.h>
+#include <c10/core/TensorTypeSet.h>
+
+namespace c10 {
+
+/**
+ * An instance of DispatchKeyExtractor knows how to get a dispatch key given
+ * a list of arguments for an operator call. The instance is specific for
+ * a certain operator as different operators have different ways to extract
+ * the dispatch key (e.g. different numbers of arguments).
+ */
+struct DispatchKeyExtractor final {
+public:
+  static DispatchKeyExtractor make(const FunctionSchema& schema) {
```

Comment:
it's easier to write this way. As a constructor, I wouldn't be able to use constructor initializer lists because there's a whole lot of code that needs to run before I know what num_args or is_valid are supposed to be.